feat(core): Implement project deletion

despairblue opened the pull request:
## Summary

This implements deleting projects. 
**Only admins are allowed to delete a project.**
**Only team projects can be deleted.**

It will:
* delete all workflows the project owns
* delete all workflow relations related to the project (sharing into and out of the project)
* delete all credentials the project owns
* delete all credential relations related to the project (sharing into and out of the project)
* delete all project relations
* delete the project

## Related tickets and issues

https://linear.app/n8n/issue/PAY-1427/delete-team-project

## Review / Merge checklist

- [x] PR title and summary are descriptive. **Remember, the title automatically goes into the changelog. Use `(no-changelog)` otherwise.** ([conventions](https://github.com/n8n-io/n8n/blob/master/.github/pull_request_title_conventions.md))
- [x] Tests included.
  > A bug is not considered fixed, unless a test is added to prevent it from happening again.
  > A feature is not complete without tests.

despairblue committed:
remove userId from SharedCredentials

despairblue committed:
update `GET /credentials/`

despairblue committed:
update `GET /credentials/new`

despairblue committed:
update `GET /credentials/:credentialId`

despairblue committed:
update `POST /credentials`

despairblue committed:
update `PATCH /credentials/:credentialId`

despairblue committed:
update `DELETE /credentials/:id`

krynble committed:
fix: Always register webhooks on startup (#8830)

despairblue committed:
update `PUT /credentials/:credentialId/share`

OlegIvaniv committed:
refactor: Update Langchain to 0.1.41 & add support for Claude 3 (#8825)

Signed-off-by: Oleg Ivaniv <me@olegivaniv.com>
Co-authored-by: Michael Kret <michael.k@radency.com>

michael-radency committed:
feat: Add more AI node info to telemetry (#8827)

despairblue committed:
update `POST /credentials/test`

elsmr committed:
feat(editor): Add sections to autocomplete dropdown (#8720)

Co-authored-by: Giulio Andreini <andreini@netseven.it>

elsmr committed:
fix(editor): Make inputs in the filter component expressions by default (#8784)

gandreini committed:
feat(editor): Improve errors in output panel (#8644)

Co-authored-by: Michael Kret <michael.k@radency.com>

hayashi-ay committed:
fix: Correct misspelling from omited to omitted (no-changelog) (#8838)

elsmr committed:
fix: Improve filter component error handling (#8832)

despairblue committed:
fixed all other tests

despairblue committed:
fix type errors and remove unused code

despairblue committed:
Merge branch 'feature/rbac' into pay-1350-make-sure-that-sharedcredentialuserid-is-not-used-anymore-to-try-2

despairblue committed:
fix tests after merge

despairblue committed:
fix lint issue

despairblue committed:
fix migration, should work on postgres

despairblue committed:
actually migrate data

despairblue committed:
escape column names correctly

despairblue committed:
remove `userId` parameter from `CredentialsHelper.getCredentials`

This is parameter is not used on the call site.

despairblue committed:
allow `global:admin`s to access external secrets

despairblue committed:
remote notes

despairblue committed:
update migration to remove `SharedWorkflow.userId`

despairblue committed:
update `POST /workflows/`

despairblue committed:
update `GET /workflows/`

despairblue committed:
update `GET /workflows/new`

despairblue committed:
update `GET /workflows/from-url`

despairblue committed:
fixup! update migration to remove `SharedWorkflow.userId`

despairblue committed:
update `GET /workflows/:workflowId`

despairblue committed:
udpate `PUT /workflows/:workflowId/share`

despairblue committed:
update `PATCH /workflows/:workflowId`

despairblue committed:
update `DELETE /workflows/:workflowId`

despairblue committed:
update `POST /workflows/run`

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
fix tests

despairblue committed:
remove `SharedWorkflowRepository.hasAccess`

despairblue committed:
fix some type errors

despairblue committed:
fixup! remove `SharedWorkflowRepository.hasAccess`

despairblue committed:
remove `SharedWorkflowRepository.findSharing`

despairblue committed:
remove `WorkflowService.isOwned` and `SharedWorkflowRepository.getSharing`

despairblue committed:
fix `SharedWorkflowRepository.makeOwnerOfAllWorkflows`

despairblue committed:
remove `SharedWorkflowRepository.share`

despairblue committed:
small type fixes

despairblue committed:
fix type issues

despairblue committed:
fix invalid relations

despairblue committed:
fix type error

despairblue committed:
remove unused parameter

despairblue committed:
Merge branch 'feature/rbac' into pay-1353-make-sure-that-sharedworkflowuserid-is-not-used-anymore-to

despairblue committed:
fix build issues

despairblue committed:
remove stray console.log

despairblue committed:
Update packages/cli/src/PublicApi/v1/handlers/workflows/workflows.handler.ts

Co-authored-by: Omar Ajoue <krynble@gmail.com>

despairblue committed:
Merge branch 'feature/rbac' into pay-1353-make-sure-that-sharedworkflowuserid-is-not-used-anymore-to

despairblue committed:
remove dead code

despairblue committed:
fix lint issue

despairblue committed:
remove comments

despairblue committed:
remove more dead code

despairblue committed:
also remove test

despairblue committed:
--wip--

despairblue committed:
Merge branch 'pay-1353-make-sure-that-sharedworkflowuserid-is-not-used-anymore-to' into pay-1427-delete-team-project

despairblue committed:
refactor project controller tests to use shared helpers

despairblue committed:
--wip--

despairblue committed:
Merge branch 'feature/rbac' into pay-1427-delete-team-project

despairblue committed:
--wip--

krynble commented on packages/cli/src/services/project.service.ts:
```diff
@@ -21,6 +26,65 @@ export class ProjectService {
+		await this.projectRelationRepository.manager.transaction(async (em) => {
+			try {
+				// 1. delete credentials owned by this project
+				const ownedCredentials = await em.find(SharedCredentials, {
+					where: { projectId: project.id, role: 'credential:owner' },
+					relations: { credentials: true },
+				});
+
+				console.log('ownedCredentials', ownedCredentials);
+
+				for (const credential of ownedCredentials) {
+					await Container.get(CredentialsService).delete(credential.credentials, em);
+				}
+
+				// 2. delete workflows owned by this project
+				const ownedSharedWorkflows = await em.find(SharedWorkflow, {
+					where: { projectId: project.id, role: 'workflow:owner' },
+					relations: { workflow: true },
+				});
+
+				for (const sharedWorkflow of ownedSharedWorkflows) {
+					await Container.get(WorkflowService).delete(user, sharedWorkflow.workflow.id, em);
+				}
+
+				// 3. delete shared credentials into this project
+				// Should cascade, but should this run the same hooks that unsharing does?
+				// 4. delete shared workflows into this project
+				// Should cascade, but should this run the same hooks that unsharing does?
+				// 5. delete project
+				await em.remove(project);
+			} catch (error) {
+				console.error(error);
+				throw error;
+			}
```
I think it's worth us switching the order - remove workflows first and then credentials.

This is simply because some workflows might be active and with webhooks.

In order to reregister the webhooks we might need the credentials used by them.

krynble commented:
Tried building in [CI](https://github.com/n8n-io/n8n/actions/runs/8298712390/job/22712696531) but failed with the errors below

![Screenshot 2024-03-15 at 15 58 54](https://github.com/n8n-io/n8n/assets/219272/b7120621-f7fd-45be-ad1b-daea52887895)

despairblue committed:
build works

despairblue committed:
delete workflows first and then credentials

despairblue committed:
add last test and clean up

despairblue committed:
fix lints

despairblue committed:
poor man's lazy dependency injection

despairblue committed:
Merge branch 'feature/rbac' into pay-1427-delete-team-project

despairblue committed:
Merge branch 'feature/rbac' into pay-1427-delete-team-project

despairblue committed:
fix integration tests and unit tests that broke due to refactors

despairblue committed:
make sure project relations are also deleted

despairblue committed:
fix lint

despairblue commented:
All those errors already exist on the feature branch. 
There are lint issues in the design system, FE unit tests failing, and the MySQL tests we already know about.

![image](https://github.com/n8n-io/n8n/assets/927609/9d6f9b9d-591b-4e26-9ee4-ce66493ee8b8)

krynble commented on packages/cli/src/workflows/workflow.service.ts:
```diff
@@ -220,31 +220,39 @@ export class WorkflowService {
 		if (workflow.active) {
 			// deactivate before deleting
-			await this.activeWorkflowRunner.remove(workflowId);
+			await this.activeWorkflowRunner.remove(workflowId, em);
```
I'm thinking whether from this point downstream we don't use the transactions and leave as regular queries, as we're getting the `entityManager` to places it's not supposed to be WDYT?

I'm ok with passing EM across repositories, but this seems to be going too far and would make it impossible to trace back where a given EM came from.

krynble replied:
WDYT @valya ?

despairblue replied:
Decision: 

`WorkflowService.delete`'s body is split out into 2 new functions:
1. deactivate workflow
	* does not take the `em`
2. delete workflow
	* does take an optional `em`

Then `ProjectService.delete` can call 1. without passing the `em` and then call 2. with passing the `em`.

That way we don't have to pass the `em` around as much. 

This leaves a race condition in the code that already exists right now: The workflow could be activated again between it being deactivated and it being deleted. 
This race condition exists already and we're ok with keeping in, given that the time between the deactivation and deletion is neglectable and it will probably never happen.

despairblue committed:
don't deactivate workflows within a transaction

despairblue committed:
allow instance owners to delete everybody's any project

despairblue committed:
fix tests that tested with wrong assumptions

it used the owner to check project permissions, but the owner will always bypass the project level permissions

despairblue committed:
fix lint issue

krynble commented on packages/cli/src/services/project.service.ts:
```diff
@@ -21,6 +24,81 @@ export class ProjectService {
+			}
+
+			// 1. delete workflows owned by this project
+			const ownedSharedWorkflows = await em.find(SharedWorkflow, {
+				where: { projectId: project.id, role: 'workflow:owner' },
+				relations: { workflow: true },
+			});
+
+			for (const sharedWorkflow of ownedSharedWorkflows) {
+				// NOTE: This is supposed to happen outside the transaction.
+				// We'd otherwise need to pass the em through to a lot of functions even ending up in core.
+				// See: https://github.com/n8n-io/n8n/pull/8904#discussion_r1530150510
+				await activeWorkflowRunner.remove(sharedWorkflow.workflow.id);
+				await workflowService.deleteInactiveWorkflow(user, sharedWorkflow.workflow, em);
+			}
```
I'd suggest moving this code to outside the transaction and only begin the transaction for the actual removal wdyt?

This way, if something fails while deactivating we abort prematurely and warn the user, and we still have the safety of the transaction.

Does this sound too crazy?

Worst case scenario we end up in a situation where the workflow was deactivated but a few remain active, which should not be too bad.

despairblue replied:
I can, but how do I do that without moving this out as well?

```ts
			const ownedSharedWorkflows = await em.find(SharedWorkflow, {
				where: { projectId: project.id, role: 'workflow:owner' },
				relations: { workflow: true },
			});
```

And if I move this out we have another race condition. A new workflow could be created while the project is being deleted.

despairblue replied:
I mean I could do the `find` twice and abort the transaction if I get different results in between.

I'm not happy with how much context a future reader will need to understand all the intricacies of this.

despairblue replied:
Nevermind, `find`s don't lock the tables, so the problem that a new workflow can be added while the project is being deleted is something that would happen even with the `find` being in the transaction.

krynble reviewed (approved):
LGTM, I just left an idea for consideration, we could deactivate all workflows as usual, before the transaction starts, and if that succeeds, we start the transaction to simply remove all the contents.

despairblue committed:
disable all workflows before the transaction starts

despairblue committed:
simplify

krynble reviewed (approved):
LGTM
